engine/access/ingestion2/results_forest_test.go

```diff
@@ -0,0 +1,1684 @@
+		container, found := forest.GetContainer(resultID)
+		s.True(found)
+		s.Equal(resultID, container.ResultID())
+		s.Equal(uint(1), container.Size())
+		s.True(container.Has(receipt.ID()))
+		s.Equal(ingestion2.ResultForCertifiedBlock, container.ResultStatus())
+
+		// this is mocked, but should be abandoned for all conflicting results
+		s.Equal(optimistic_sync.StateAbandoned, container.Pipeline().GetState())
+	}
+}
+
+// TestAddReceipt tests adding execution receipts to the forest.
+// It covers successful addition, duplicate handling, and error cases.
+func (s *ResultsForestSuite) TestAddReceipt() {
+	resultGen := s.g.ExecutionResults()
+	receiptGen := s.g.ExecutionReceipts()
+
+	s.Run("add receipt for new result that extends an existing result", func() {
+		forest := s.createForest()
+
+		result := s.results[0]
+		receipt := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+
+		pipeline := s.createPipeline(result)
+		pipeline.On("SetSealed").Return().Once()
+
+		added, err := forest.AddReceipt(receipt, ingestion2.ResultSealed)
+		s.Require().NoError(err)
+		s.True(added)
+
+		s.assertContainer(forest, result.ID(), ingestion2.ResultSealed)
+		s.Equal(uint(2), forest.Size())
+	})
+
+	s.Run("add receipt for new result that does not extend an existing result", func() {
+		forest := s.createForest()
+
+		result := s.results[1]
+		receipt := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+
+		_ = s.createPipeline(result)
+
+		added, err := forest.AddReceipt(receipt, ingestion2.ResultForCertifiedBlock)
+		s.Require().NoError(err)
+		s.True(added)
+
+		s.assertContainer(forest, result.ID(), ingestion2.ResultForCertifiedBlock)
+		s.Equal(uint(2), forest.Size())
+	})
+
+	s.Run("add receipt for existing result", func() {
+		forest := s.createForest()
+
+		result := s.results[0]
+		receipt1 := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+		receipt2 := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+
+		_ = s.createPipeline(result)
+
+		// add the first receipt
+		added, err := forest.AddReceipt(receipt1, ingestion2.ResultForCertifiedBlock)
+		s.Require().NoError(err)
+		s.True(added)
+
+		// result should be added to the forest
+		s.assertContainer(forest, result.ID(), ingestion2.ResultForCertifiedBlock)
+		s.Equal(uint(2), forest.Size())
+
+		// add the second receipt
+		added, err = forest.AddReceipt(receipt2, ingestion2.ResultForCertifiedBlock)
+		s.Require().NoError(err)
+		s.True(added)
+	})
+
+	s.Run("add receipt for finalized result with view lower that finalized view", func() {
+		forest := s.createForest()
+
+		result := s.results[0]
+		receipt := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+
+		// progress finalized view to block[0]
+		err := forest.OnBlockFinalized(s.blocks[0])
+		s.Require().NoError(err)
+
+		_ = s.createPipeline(result)
+
+		added, err := forest.AddReceipt(receipt, ingestion2.ResultForFinalizedBlock)
+		s.Require().NoError(err)
+		s.True(added)
```

Comment:
maybe also check the size of the container? It's size should have increased, right?